src/dinitctl.cc

```diff
@@ -1903,39 +1903,61 @@ static int enable_disable_service(int socknum, cpbuffer_t &rbuffer, service_dir_
     }
 
     if (socknum >= 0) {
+        if (verbose) {
+            cout << "Service '" << to << "' has been " << (enable ? "enabled" : "disabled") << "." << endl;
+        }
+
         // Check status of the service now
         auto m = membuf()
                 .append((char)cp_cmd::SERVICESTATUS)
                 .append(to_handle);
         write_all_x(socknum, m);
 
-        wait_for_reply(rbuffer, socknum);
-        if (rbuffer[0] != (char)cp_rply::SERVICESTATUS) {
+        int r = rbuffer.fill_to(socknum, 2);
+        while (r > 0 && rbuffer[0] >= 100) {
+            unsigned pktlen = (unsigned char) rbuffer[1];
+            fill_buffer_to(rbuffer, socknum, pktlen);
+
+            if (rbuffer[0] == (char)cp_info::SERVICEEVENT) {
+                wait_service_state(socknum, rbuffer, to_handle, to, !enable, verbose);
+                return 0;
```

Comment:
Bailing out here leaves the `SERVICESTATUS` reply still pending. In general the functions in dinitctl do not leave the communication pipe in a non-neutral state (if they return success). The reply should be consumed before returning. Not doing so could lead to bugs later if callers assume that the pipe will be in a neutral state (i.e. no replies pending).